normalizer.go

```diff
@@ -0,0 +1,219 @@
+	Commands []string
+}
+
+type Normalizer struct {
+	config *normalizerConfig
+}
+
+func NewNormalizer(opts ...func(*normalizerConfig)) *Normalizer {
+	normalizer := Normalizer{
+		config: &normalizerConfig{},
+	}
+
+	for _, opt := range opts {
+		opt(normalizer.config)
+	}
+
+	return &normalizer
+}
+
+const (
+	ArrayPlaceholder   = "( ? )"
+	BracketPlaceholder = "[ ? ]"
+)
+
+// Normalize takes an input SQL string and returns a normalized SQL string, a StatementMetadata struct, and an error.
+// The normalizer collapses input SQL into compact format, groups obfuscated values into single placeholder,
+// and collects metadata such as table names, comments, and commands.
+func (n *Normalizer) Normalize(input string) (normalized string, info *StatementMetadata, err error) {
+	lexer := New(input)
+
+	var normalizedSQL string
+	var statementMetadata = &StatementMetadata{
+		Tables:   []string{},
+		Comments: []string{},
+		Commands: []string{},
+	}
+
+	var lastToken *Token // The last token that is not whitespace or comment
+
+	for token := range lexer.ScanAllTokens() {
+		if token.Type == COMMENT || token.Type == MULTILINE_COMMENT {
+			// Collect comments
+			if n.config.CollectComments {
+				statementMetadata.Comments = append(statementMetadata.Comments, token.Value)
+			}
+		} else if token.Type == IDENT {
+			if isCommand(strings.ToUpper(token.Value)) && n.config.CollectCommands {
+				// Collect commands
+				statementMetadata.Commands = append(statementMetadata.Commands, strings.ToUpper(token.Value))
+			} else if lastToken != nil && isTableIndicator(strings.ToUpper(lastToken.Value)) {
+				// Collect table names
+				if n.config.CollectTables {
+					statementMetadata.Tables = append(statementMetadata.Tables, token.Value)
+				}
+			}
+		}
+
+		normalizedSQL = writeNormalizedSQL(token, lastToken, normalizedSQL)
+
+		// TODO: We rely on the WS token to determine if we should add a whitespace
+		// This is not ideal, as SQLs with slightly different formatting will NOT be normalized into single family
+		// e.g. "SELECT * FROM table where id = ?" and "SELECT * FROM table where id= ?" will be normalized into different family
+		if token.Type != WS && token.Type != COMMENT && token.Type != MULTILINE_COMMENT {
+			lastToken = token
+		}
+	}
+
+	// We use regex to group consecutive obfuscated values into single placeholder.
+	// This is "less" performant than token by token processing,
+	// but it is much simpler to implement and maintain.
+	// The trade off made here is assuming normalization runs on backend
+	// where performance is not as critical as the agent.
+	normalizedSQL = groupObfuscatedValues(normalizedSQL)
+	if !n.config.KeepSQLAlias {
+		normalizedSQL = DiscardSQLAlias(normalizedSQL)
+	}
+
+	// Dedupe collected metadata
+	dedupeStatementMetadata(statementMetadata)
+
+	return strings.TrimSpace(normalizedSQL), statementMetadata, nil
+}
+
+func writeNormalizedSQL(token *Token, lastToken *Token, statement string) string {
```

Comment:
I think that now that this is an input -> output function, it might be better named something like `normalizeSQL`?